rmw_opensplice_cpp/src/functions.cpp

```diff
@@ -878,10 +878,17 @@ rmw_wait(
 
   // invoke wait until one of the conditions triggers
   DDS::ConditionSeq active_conditions;
+  bool non_blocking = false;
   DDS::Duration_t timeout;
-  timeout.sec = non_blocking ? 0 : 1;
-  timeout.nanosec = 0;
+  if (!wait_timeout) {
+    timeout = DDS::DURATION_INFINITE;
+  } else {
+    non_blocking = wait_timeout->sec == 0 && wait_timeout->nsec == 0;
+    timeout.sec = wait_timeout->sec;
+    timeout.nanosec = wait_timeout->nsec;
+  }
   DDS::ReturnCode_t status = DDS::RETCODE_TIMEOUT;
+
   while (DDS::RETCODE_TIMEOUT == status) {
     status = waitset.wait(active_conditions, timeout);
     if (DDS::RETCODE_TIMEOUT == status) {
```

Comment:
As I was writing a test for this, I realized that the existing logic of this loop is not what I thought it was.

Old behavior:

* Get "blocking" or "non-blocking in rclcpp.
* In rmw implementation:

* if non-blocking, call rmw_wait with a timeout of zero, and don't loop if timed out.
* if blocking: while return code is timeout, wait for 1 second. throw if error, continue if retcode was not timeout. The executor will always block indefinitely inside of rmw_wait.

Here's what I think the new behavior should be:

Get timeout from rclcpp.
* If NULL: block indefinitely in rmw_wait, by passing `DURATION_INFINITE` to `WaitSet::Wait`.
* If positive: pass the timeout value to `WaitSet::wait` and return if timed out.
* If non-blocking: same behavior as before.

Thoughts on this? (@wjwood?)